### DISCUSSION TITLE
LTS awareness

### DISCUSSION THREAD

**Discussion by iamstarkov** [renovatebot/renovate#17082] [Request Help] [answered]
I read through documentation and I couldn't find LTS awareness option. As in, I'd like to use latest nodejs/java/etc but only if it is an LTS release.

For example, consider [Node.JS releases](https://github.com/nodejs/Release)
![schedule](https://user-images.githubusercontent.com/559321/183691249-27ec1e84-a8e8-4ceb-8621-3993f1d3216e.svg)

As of now, latest LTS is 16. Although version 18 is available but it wasn't promoted to LTS just yet, which is bound to happen in 2022-10-25.

Let's say, there is an configuration option `lts-active-latest`, then if my current node is `16.8.0`, then immediate update will be to `16.16.0` and on 25 October there will be a bump to `18.X.X`.

What do you think of LTS award updates?

edit: answer is that renovate does respect node LTS policy if its from node data source, but node in docker context doesnt equal to node data source renovate doesn't have much of a choice but to create an update for the latest node:X.X.X-alpine

**Comment by Gabriel-Ladzaretti** [comment, marked as answer]
currently renovate will filter out non LTS for nodejs
https://docs.renovatebot.com/node/
can be overriden with `ignoreUnstable` - https://docs.renovatebot.com/configuration-options/#ignoreunstable

**Reply by Gabriel-Ladzaretti** [reply]
here is the schedule renovate uses for nodejs
https://github.com/renovatebot/renovate/blob/5069b6010933e3aab044e703ec49e4a2d7d8a9ac/data/node-js-schedule.json#L86-L104

**Reply by Gabriel-Ladzaretti** [reply]
so renovate currently will offer updates to v16 only, as it is the latest active LTS.

**Reply by iamstarkov** [reply]
im using default config.

im not getting v17 as intended, but I do get an update to v18 (not intended) pull-request while its not and LTS just yet and only will become [LTS on `"lts": "2022-10-25"`](https://github.com/renovatebot/renovate/blob/5069b6010933e3aab044e703ec49e4a2d7d8a9ac/data/node-js-schedule.json#L100)


is it intentional?

**Reply by viceice** [reply]
your probably not using node versioning, please provide more details about the pr

**Reply by iamstarkov** [reply]
@Gabriel-Ladzaretti so like im with you on renovate filtering out non-lts, but I think its jumping the gun over the even releases which are yet to become lts

**Reply by iamstarkov** [reply]
pull-request
![Screenshot 2022-08-10 at 11 20 41](https://user-images.githubusercontent.com/559321/183865171-b1fe9a96-03fb-4149-b675-0d4934409da8.png)

config
![Screenshot 2022-08-10 at 11 22 39](https://user-images.githubusercontent.com/559321/183865574-6a3322bd-5550-457d-b133-fdc59e6974d2.png)

**Reply by Gabriel-Ladzaretti** [reply]
If i'm not mistaken and by the looks of it you are using the docker datasource (alpine suffix) for this specific PR update. therefore the versioning used is not the node related one.
~~try adding a package rule for the node package and the docker manager and set the versioning to node.~~

**Reply by rarkins** [reply]
Yeah we don't get support both LTS awareness and platform/suffix awareness in one

**Reply by viceice** [reply]
https://github.com/renovate-reproductions/16925-docker-digest-update-group-test/pull/5
it seems something is still wrong, there shouldn't be an update to node v18

**Reply by viceice** [reply]
reproduction : https://github.com/renovate-reproductions/docker-node-versioning-issue/pull/2

**Reply by viceice** [reply]
```json
       "config": {
         "dockerfile": [
           {
             "packageFile": "Dockerfile",
             "deps": [
               {
                 "depName": "node",
                 "currentValue": "16.15.3",
                 "replaceString": "node:16.15.3",
                 "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest}}@{{newDigest}}{{/if}}",
                 "datasource": "docker",
                 "depType": "final",
                 "depIndex": 0,
                 "updates": [
                   {
                     "bucket": "minor",
                     "newVersion": "16.16.0",
                     "newValue": "16.16.0",
                     "newMajor": 16,
                     "newMinor": 16,
                     "updateType": "minor",
                     "branchName": "renovate-test/node-16.x"
                   },
                   {
                     "bucket": "major",
                     "newVersion": "18.7.0",
                     "newValue": "18.7.0",
                     "newMajor": 18,
                     "newMinor": 7,
                     "updateType": "major",
                     "branchName": "renovate-test/node-18.x"
                   }
                 ],
                 "warnings": [],
                 "versioning": "docker",
                 "sourceUrl": "https://github.com/nodejs/node",
                 "currentVersion": "16.15.3",
                 "isSingleVersion": true,
                 "fixedVersion": "16.15.3"
               }
             ]
           }
         ]
       }
```

**Reply by Gabriel-Ladzaretti** [reply]
the docker versioning is used in the case of `node:16.8.0-alpine` for example

<details><summary>package file with updates log for docker dep</summary>

```
{
             "packageFile": "Dockerfile",
             "deps": [
               {
                 "depName": "node",
                 "currentValue": "16.8.0-alpine",
                 "replaceString": "node:16.8.0-alpine",
                 "autoReplaceStringTemplate": "{{depName}}{{#if newValue}}:{{newValue}}{{/if}}{{#if newDigest
}}@{{newDigest}}{{/if}}",
                 "datasource": "docker",
                 "depType": "final",
                 "depIndex": 0,
                 "updates": [
                   {
                     "bucket": "non-major",
                     "newVersion": "16.16.0",
                     "newValue": "16.16.0-alpine",
                     "newMajor": 16,
                     "newMinor": 16,
                     "updateType": "minor",
                     "branchName": "renovate/node-16.x"
                   },
                   {
                     "bucket": "major",
                     "newVersion": "18.7.0",
                     "newValue": "18.7.0-alpine",
                     "newMajor": 18,
                     "newMinor": 7,
                     "updateType": "major",
                     "branchName": "renovate/node-18.x"
                   }
                 ],
                 "warnings": [],
                 "versioning": "docker",
                 "sourceUrl": "https://github.com/nodejs/node",
                 "currentVersion": "16.8.0",
                 "isSingleVersion": true,
                 "fixedVersion": "16.8.0-alpine"
               }

```

</details>

**Reply by Gabriel-Ladzaretti** [reply]
i also confirmed it locally, but does the docker versioning should be aware of the node scheduling as the node versioning does?

**Reply by viceice** [reply]
no, sorry, may fault, i forgot about the docker issue 😕

**Reply by Gabriel-Ladzaretti** [reply]
related to this, I'm wondering is it possible to override versioning as part of a package rule? i wasnt able to find such an option and it seems that it is only possible as part of a regex manager.

second question, if for once i define a regex manager to handle a manifest file that is already handled out of the box (e.g. dockerfile). will the default manager be disabled, or is it theoretically possible for two managers to operate on the same file? 

Above is basically a build up for this as a solution: https://docs.renovatebot.com/modules/manager/regex/#advanced-capture this has a regex manager that can set versioning dynamically to a dockerfile dep, i thought to define such a manager to handle `node` only and use the default one for all other dockerfile related deps.

**Reply by viceice** [reply]
you can override versioning via package rules

**Reply by nabeelsaabna** [reply]
```
	"packageRules": [
		{
			"matchPackageNames": [
				"package-name"
			],
			"versioning": "semver"
		}
	]
```

**Reply by Gabriel-Ladzaretti** [reply]
thanks @nabeelsaabna @viceice 
managed to find it https://docs.renovatebot.com/configuration-options/#versioning

**Reply by iamstarkov** [reply]
> i also confirmed it locally, but does the docker versioning should be aware of the node scheduling as the node versioning does?

I'd reckon yes, if its one of the few official [nodejs docker distributions](https://hub.docker.com/_/node/)

**Reply by iamstarkov** [reply]
@nabeelsaabna @Gabriel-Ladzaretti should not I use `node` value for versioning for `node` package instead of `semver` because its more relevant?

```
	"packageRules": [
		{
			"matchPackageNames": [
				"node"
			],
			"versioning": "node"
		}
	]
```

![Screenshot 2022-08-11 at 16 07 19](https://user-images.githubusercontent.com/559321/184152470-362421ae-268d-4739-bd1d-ff00538eddd8.png)


also wouldn't it cause docker/distribution specific bits (`-alpine`) will be discarded and renovate update will become less specific than it needs to be:
* before: `16.8.0-alpine -> 18.7.0-alpine`
* after: `16.8.0-alpine -> 16.6.0`

like im glad it were to be within LTS range, but I still need distribution specific.

What do you think?

**Reply by rarkins** [reply]
While you have the `-alpine` suffix, you can only really use `docker` versioning

**Reply by iamstarkov** [reply]
oh, its a shame

**Reply by viceice** [reply]
@rarkins i think we can make the suffix for node versioning a compabillity value, as node has never released prereleases?

**Reply by rarkins** [reply]
See https://github.com/renovatebot/renovate/issues/5990

In this case instead of having "docker" versioning - which is essentially loose semver versioning combined with compatibility suffixes - we could use the concept of `versionCompatibility` to teach Renovate which part is the version and which is the compatibility. Then any `versioning` can be used for the `version` part, including `node` versioning. `docker` versioning would essentially be unncessary.

**Reply by iamstarkov** [reply]
@Gabriel-Ladzaretti @rarkins @viceice thank you for all of your help, I've marked this reply as an answer and also edited an question to have a more detailed answer from my point of view

**Comment by rarkins** [comment]
Which package manager/use cases do you need for the `node` case? So far we are solving this with custom versioning when an ecosystem is large enough (e.g. node, ubuntu, debian)

**Reply by iamstarkov** [reply]
@rarkins im not quite sure if I understand your question in its entirety, but I'll try my best. We are using docker, npm, yarn, PNPM, nvm and `package.json` engines to align and coordinate nodejs version between local dev machines, CI and production environments
